### DISCUSSION TITLE
Can I create *. model file from checkpoint file?

### DISCUSSION THREAD

**Discussion by turbosonics** [ACEsuit/mace#564] [Q&A] [answered]
Hi, 

I have some training jobs which are over by wall-time limit before reaching target epoch. In this case, because the training job is killed, output folder didn't create *_compiled/model file or *.model file. 

But I hope to test at least some of these training result from LAMMPS MD simulation before I restart this training work. 

So I'm curious, how can I generate these model files from checkpoint pt file? Once I have model file then I can convert it to lammps-ready pt file using create_lammps_model.py. 

Or are there any other ways to generate lammps-ready model file from training jobs that is finished before reaching target epoch? 

Thanks

**Comment by ilyes319** [comment, marked as answer]
There is no direct wait to do that yet, but the easiest is the following:
 - First run a short job, to create a temple *.model that has the same hypers as the final mode.
 - Then use a following script:
 
```
ckpt = torch.load(path_ckpt)
model = torch.load(path_model)
model.load_state_dict(ckpt["model"])
torch.save(model, path_new_model)
```

**Reply by legend-L24** [reply]
Hello, I tried this code but met errors. The code and the error are shown below. Do you have any advice?

Code:
"""

I fine-tune the MACE-MP-0b2 model. And MACE_run-3_epoch-150.pt comes from checkpoints during a long training. MACE_compiled.model is a saved model from a short training. And I want to load a model from MACE_run-3_epoch-150.pt. But the output model does not contain __get_state function.

"""



import torch

path_ckpt = "/data2/yutli/mace_0303/checkpoints/MACE_run-3_epoch-150.pt"
path_model = "/data2/yutli/mace_0303/MACE_compiled.model"
path_new_model = "/data2/yutli/mace_0303/mace-finish.model"
ckpt = torch.load(path_ckpt, weights_only=False)
model = torch.load(path_model, weights_only=False)
model.load_state_dict(ckpt["model"])
torch.save(model, path_new_model)

Error:

RuntimeError                              Traceback (most recent call last)
Cell In[25], line 9
      7 model = torch.load(path_model, weights_only=False)
      8 model.load_state_dict(ckpt["model"])
----> 9 torch.save(model, path_new_model)

File ~/software/anaconda3/envs/gnnpt/lib/python3.12/site-packages/torch/serialization.py:944, in save(obj, f, pickle_module, pickle_protocol, _use_new_zipfile_serialization, _disable_byteorder_record)
    942 if _use_new_zipfile_serialization:
    943     with _open_zipfile_writer(f) as opened_zipfile:
--> 944         _save(
    945             obj,
    946             opened_zipfile,
    947             pickle_module,
    948             pickle_protocol,
    949             _disable_byteorder_record,
    950         )
    951         return
    952 else:

File ~/software/anaconda3/envs/gnnpt/lib/python3.12/site-packages/torch/serialization.py:1190, in _save(obj, zip_file, pickle_module, pickle_protocol, _disable_byteorder_record)
   1187         return persistent_id(obj)
   1189 pickler = PyTorchPickler(data_buf, protocol=pickle_protocol)
-> 1190 pickler.dump(obj)
   1191 data_value = data_buf.getvalue()
   1192 zip_file.write_record("data.pkl", data_value, len(data_value))

RuntimeError: Tried to serialize object __torch__.mace.modules.models.ScaleShiftMACE which does not have a __getstate__ method defined!

**Reply by ilyes319** [reply]
Hey, you should not use the model with "compiled" in it, that is the JIT compiled version. Please try with the other saved model.

**Comment by pdobbelaere** [comment]
Hi,
I know this is an older issue, but have there been any developments regarding this functionality in recent MACE versions?
It is quite a bother to launch a dummy training job to build the torch model if something happens to the actual training run.
Perhaps the model architecture can be stored directly after initialisation, just in case ([here](https://github.com/ACEsuit/mace/blob/42cd495b9c8e155441c405fef842a8ba5107c2f8/mace/cli/run_train.py#L686) somewhere)?

Related question: I do not see any `avg_num_neighbors` keys in the state_dict, so I assume that value is stored as a python instance attribute rather than a torch buffer? If so, that might cause unexpected behaviour (because e.g., `r_max` is in the state_dict).

**Reply by pdobbelaere** [reply]
friendly bump
@ilyes319
